feat: make send_event assign the target step

Tonyrj3268 opened the pull request:
# Description

Add step parameter to let step_event method can assign the target step.
If step is None, it will run like the origin one.
If the target step is not existed or the step does not accept this event type, it will raise WorkflowValidationError.

Fixes # (issue)

## New Package?

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [ ] Yes
- [x] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [ ] Yes
- [x] No

## Type of Change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

## Suggested Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

nerdai reviewed (approved):
This looks pretty good to me, @Tonyrj3268!

Going to request @masci's eyes on this too as I'm still getting up to speed with our Workflows

masci commented on llama-index-core/llama_index/core/workflow/workflow.py:
```diff
@@ -157,15 +157,29 @@ async def _task(
+                queue.put_nowait(message)
+
+        if step not in self._get_steps():
+            raise WorkflowValidationError(f"Step {step} does not exist")
```
nit: `send_event` is not called during validation, this should probably be some sort of runtime error

masci commented on llama-index-core/llama_index/core/workflow/workflow.py:
```diff
@@ -157,15 +157,29 @@ async def _task(
+        if step is None:
+            for queue in self._queues.values():
+                queue.put_nowait(message)
+
```
I think we should return early here, or put the rest in an else block, otherwise `step not in self._get_steps()` below would fail I think. I would also add a unit test to cover specifically when step is `None` (tests fail locally for me indeed).

masci commented on llama-index-core/llama_index/core/workflow/workflow.py:
```diff
@@ -157,15 +157,29 @@ async def _task(
+        if step_config and type(message) in step_config.accepted_events:
+            self._queues[step].put_nowait(message)
+        else:
+            raise WorkflowValidationError(
```
Same consideration about the error type

masci reviewed (changes requested):
Thanks for tackling this!

I left a couple of comments, but regarding the overall design I am not sure we need to go always through the same logic to get the list of accepted events. One alternative might be building a data structure mapping step->List[supported_events] once, and accessing it super cheaply at every `send_event`.

Tonyrj3268 commented:
@masci 
Thank you for your review. I have corrected the exception type and added an else block (I didn't return early because we need to record the log).
Additionally, I strongly agree about having a data structure that maps steps to events. The current process of finding the function and then looking for accepted events in its config is weird.
But, I'm not quite sure how llama_index plan to implement this mapping and the other feature requests you mentioned in the issue. So I haven't taken any next steps for workflow.

masci reviewed (approved):
LGTM! We can work it out a better API to map steps to accepted events later.
